CRM/Upgrade/Incremental/php/FourSeven.php

```diff
@@ -728,6 +738,37 @@ public function upgradeMailingFKs(CRM_Queue_TaskContext $ctx) {
   }
 
   /**
+   * CRM-17663 - Dashboard schema changes
+   *
+   * @param \CRM_Queue_TaskContext $ctx
+   *
+   * @return bool
+   */
+  public function dashboardSchemaUpdate(CRM_Queue_TaskContext $ctx) {
+    if (!CRM_Core_BAO_SchemaHandler::checkIfIndexExists('civicrm_dashboard_contact', 'index_dashboard_id_contact_id')) {
+      CRM_Core_DAO::executeQuery('ALTER TABLE civicrm_dashboard_contact ADD UNIQUE INDEX index_dashboard_id_contact_id (dashboard_id, contact_id);');
```

Comment:
@colemanw hmm good point happy to leave the uniqueness as is, it was just something i was working on as a theory when trying to get my test db to behave. I guess if its not supported by the GUI then we shouldn't support it in the database. 

I think we should assume there is dupes around and get the database into a correct state, For example look at the work we had to do with the settings conversion further up the file